tools/purge_ems.rb

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env ruby
+require File.expand_path('../config/environment', __dir__)
+require 'optimist'
+
+ARGV.shift if ARGV[0] == '--' # if invoked with rails runner
+batch_message = "For each relationship, the number of rows to be removed in a batch. A lower number may decrease memory usage but also take longer."
+opts = Optimist.options do
+  banner "Purge management system records in the background via the queue.\n\nUsage: ruby #{$PROGRAM_NAME} [options]\n\nOptions:\n\t"
+  opt :id,      "Management System id",                                         :type => :integer,                    :required => true
+  opt :follow,  "Follow the progress or exit after creating work items.",       :type => :boolean, :default => true
+  opt :batch,   batch_message,                                                  :type => :integer, :default => 1_000
+  opt :timeout, "The number of minutes each work item will be allowed to run.", :type => :integer, :default => 30
+end
+
+Optimist.die :id,      "must be a positive number"                             if opts[:id] < 1
+Optimist.die :batch,   "must be a positive number"                             if opts[:batch] < 1
+Optimist.die :timeout, "must be a positive number greater than or equal to 10" if opts[:timeout] < 10
+
+DESTROY_METHOD = "destroy".freeze
+ZONE           = nil                    # Will queue for any zone
+PRIORITY       = MiqQueue::LOW_PRIORITY # Don't block other work
+
+def log(msg)
+  $log.info("MIQ(#{__FILE__}) #{msg}")
+  puts msg
+end
+
+def current_backlog
+  MiqQueue.where(:method_name => DESTROY_METHOD, :zone => ZONE, :priority => PRIORITY).count
+end
+
+id = opts[:id]
+ems = ExtManagementSystem.find(id)
+if ems.enabled?
+  log("Pausing management system with id: #{id} #{ems.name}...")
+  ems.pause!
+  log("Pausing management system with id: #{ems.id} #{ems.name}...Complete")
+end
+
+log("Removing any workers for management system with id: #{id} #{ems.name}...")
+ems.ems_workers.each(&:kill_async)
+ems.wait_for_ems_workers_removal
+log("Removing any workers for management system with id: #{ems.id} #{ems.name}...Complete")
+
+log("Removing any child managers for management system with id: #{id} #{ems.name}...")
+ems.child_managers.each(&:orchestrate_destroy)
```

Comment:
Doesn't this just reintroduce the same problems that orchestrate_destroy has?  i.e. if a child manager's orchestrate_destroy is slow, then this will just get stuck?

Reply:
Agreed, if anything I think we'd want to do this faster dependent destroy method but on the child managers.

Reply:
Ok, I can also do the child managers.

Reply:
Ok, I pushed a commit (ignore whitespace to see it) to create methods for the work needed for a single ems and a top level method to drive the removal of the top level ems and it's children.

Reply:
I'll update screenshots when the dust settles ;-)

Reply:
ok, updated screenshots and applied suggestions